File: vt_prefilter/vt_prefilter.go
package vt_prefilter

import (
	"math"
)

// VTConstants for Varshamov-Tenengolts encoding
const (
	VTAlphabet = "abcdefghijklmnopqrstuvwxyz"
	VTSigma    = len(VTAlphabet)
	VTM        = 67 // ≥ sigma+1, prime
)

// vtToInt maps characters to their 1-based index
var vtToInt = make(map[rune]int)

func init() {
	for i, c := range VTAlphabet {
		vtToInt[c] = i + 1
	}
}

// posMod returns a positive modulo result (like Python's % operator)
func posMod(a, m int) int {
	return ((a % m) + m) % m
}

// VTKey represents a Varshamov-Tenengolts key (n, S, P)
type VTKey struct {
	N int // length
	S int // sum of (i+1)*char_value
	P int // sum of char_values mod M
}

// CalculateVTKey calculates the VT key for a word
func CalculateVTKey(word string) VTKey {
	n := len(word)
	if n == 0 {
		return VTKey{0, 0, 0}
	}

	var s, p int
	for i, ch := range word {
		val := vtToInt[ch]
		s += (i + 1) * val // (i+1) because we want 1-based indexing
		p += val
	}
	return VTKey{
		N: n,
		S: posMod(s, n+1), // mod (n+1) because we want values in [0,n]
		P: posMod(p, VTM), // mod M to keep values in [0,M-1]
	}
}

// VTEdits1 generates all distance-1 string variants
func VTEdits1(w string) map[string]bool {
	result := make(map[string]bool)

	// Deletions
	for i := 0; i < len(w); i++ {
		result[w[:i]+w[i+1:]] = true
	}

	// Substitutions
	for i := 0; i < len(w); i++ {
		for _, c := range VTAlphabet {
			if rune(w[i]) != c {
				result[w[:i]+string(c)+w[i+1:]] = true
			}
		}
	}

	// Insertions
	for i := 0; i <= len(w); i++ {
		for _, c := range VTAlphabet {
			result[w[:i]+string(c)+w[i:]] = true
		}
	}

	return result
}

// VTEdits1Damerau includes adjacent transpositions
func VTEdits1Damerau(w string) map[string]bool {
	result := VTEdits1(w)

	// Add adjacent transpositions
	for i := 0; i < len(w)-1; i++ {
		swapped := w[:i] + string(w[i+1]) + string(w[i]) + w[i+2:]
		result[swapped] = true
	}

	return result
}

// VTVariants returns keys of all words at Damerau-Levenshtein distance ≤ 1
func VTVariants(word string) map[VTKey]bool {
	n := len(word)
	if n == 0 {
		// For empty string, consider all possible single-character insertions
		result := map[VTKey]bool{{0, 0, 0}: true} // Include empty key
		for _, c := range VTAlphabet {
			result[CalculateVTKey(string(c))] = true
		}
		return result
	}

	// Calculate original key
	key := CalculateVTKey(word)
	result := map[VTKey]bool{key: true}

	// Get all variants using existing function
	variants := VTEdits1Damerau(word)

	// Calculate VT keys for each variant
	for variant := range variants {
		result[CalculateVTKey(variant)] = true
	}

	return result
}

// VTVariantsK returns keys reachable by ≤ k Damerau-Levenshtein edits
func VTVariantsK(word string, k int) map[VTKey]bool {
	seenWords := map[string]bool{word: true}
	frontier := map[string]bool{word: true}
	keys := make(map[VTKey]bool)

	// Include original word's key
	keys[CalculateVTKey(word)] = true

	for i := 0; i < k; i++ {
		nextFrontier := make(map[string]bool)
		for w := range frontier {
			// Generate next layer variants
			variants := VTEdits1Damerau(w)
			for v := range variants {
				if !seenWords[v] {
					seenWords[v] = true
					nextFrontier[v] = true
					// Calculate VT key for each new variant
					keys[CalculateVTKey(v)] = true
				}
			}
		}
		frontier = nextFrontier
		if len(frontier) == 0 {
			break
		}
	}

	return keys
}

// VTBound returns the theoretical upper bound on variant count for k edits
func VTBound(n, k int) int {
	if n == 0 {
		if k == 0 {
			return 1 // Only empty string
		}
		return VTSigma + 1 // Empty string + all single character insertions
	}

	if k == 0 {
		return 1 // Only the original word
	}

	// For k=1, calculate bound mathematically
	if k == 1 {
		// For a word of length n:
		// - Original word: 1
		// - Deletions: Each deletion position can result in different VT keys
		//   because the remaining characters' positions change.
		//   For each position i (0 to n-1):
		//   - Characters after i shift left
		//   - Affects both S and P values differently based on position
		//   So we need to consider each position and each possible word:
		//   n positions × VTSigma possible words = n×VTSigma
		// - Substitutions: Each position and character can result in different VT keys
		//   For each position i (0 to n-1) and each character c:
		//   - Changes both S and P values differently based on position
		//   So we need to consider each position and character:
		//   n positions × (VTSigma-1) characters × VTSigma possible words = n×(VTSigma-1)×VTSigma
		// - Insertions: Each position can result in different VT keys
		//   For each position i (0 to n), inserting a character c:
		//   - Changes the position of all characters after i
		//   - Affects both S and P values differently based on position
		//   So we need to consider each position separately:
		//   (n+1) positions × VTSigma characters × VTSigma possible words = (n+1)×VTSigma×VTSigma
		// - Transpositions: Each transposition can result in different VT keys
		//   For each position i (0 to n-2):
		//   - Changes positions of two characters
		//   - Affects both S and P values differently based on position
		//   So we need to consider each position and each possible word:
		//   (n-1) positions × VTSigma possible words = (n-1)×VTSigma
		deletions := n * VTSigma
		substitutions := n * (VTSigma - 1) * VTSigma
		insertions := (n + 1) * VTSigma * VTSigma
		transpositions := max(0, n-1) * VTSigma
		return 1 + deletions + substitutions + insertions + transpositions
	}

	// For k>1, calculate bound by considering all possible intermediate lengths
	// and the maximum number of variants possible at each length
	maxLen := n + k       // Maximum possible length after k insertions
	minLen := max(0, n-k) // Minimum possible length after k deletions

	// Calculate maximum variants at each possible length
	total := 0
	for l := minLen; l <= maxLen; l++ {
		// At each length l, we can have:
		// - Original word: 1
		// - l×VTSigma deletions
		// - l×(VTSigma-1)×VTSigma substitutions
		// - (l+1)×VTSigma×VTSigma insertions
		// - (l-1)×VTSigma transpositions (if l >= 2)
		variantsAtLength := 1 + // original word
			l*VTSigma + // deletions
			l*(VTSigma-1)*VTSigma + // substitutions
			(l+1)*VTSigma*VTSigma + // insertions
			max(0, l-1)*VTSigma // transpositions

		// For k>1, we need to consider that each edit can be applied to any position
		// in any intermediate string. The number of possible intermediate strings
		// grows exponentially with k, so we multiply by k^k to account for this.
		// This is a very loose bound, but it's guaranteed to be larger than the
		// actual number of variants.
		total += variantsAtLength * int(math.Pow(float64(k), float64(k)))
	}

	return total
}

File: vt_prefilter/vt_prefilter.py
import random, itertools

# --- constants ----------------------------------------------------------
ALPHABET = "abcdefghijklmnopqrstuvwxyz"
sigma    = len(ALPHABET)
M        = 67                               # ≥ sigma+1, prime
to_int   = {c: i+1 for i, c in enumerate(ALPHABET)}

# --- VT key -------------------------------------------------------------
def vt_key(word: str):
    n = len(word)
    if n == 0: return (0, 0, 0)
    S = sum((i+1)*to_int[ch] for i, ch in enumerate(word)) % (n+1)
    P = sum(to_int[ch] for ch in word) % M
    return (n, S, P)

# --- enumerate *every* distance‑1 string variant ------------------------
def edits1(w: str):
    splits  = [(w[:i], w[i:]) for i in range(len(w)+1)]
    deletes = [L+R[1:]                 for L, R in splits if R]
    repls   = [L+c+R[1:] for L, R in splits if R for c in ALPHABET if c!=R[0]]
    inserts = [L+c+R     for L, R in splits for c in ALPHABET]
    return set(deletes + repls + inserts)

def edits1_damerau(w: str):
    """all distance‑1 words, incl. adjacent transpositions"""
    base = edits1(w)                         # ins / del / sub (old)
    swaps = {w[:i] + w[i+1] + w[i] + w[i+2:]
             for i in range(len(w)-1)}
    return base | swaps

# --- algebraic variant generator (k=1) ----------------------------------
# --- distance‑1 variants (now with transpositions) ----------------------
def vt_variants(word: str):
    """
    Keys of all words at Damerau‑Levenshtein distance ≤ 1
    (substitution · insertion · deletion · adjacent‑swap).
    """
    n = len(word)
    if n == 0:
        return {(0, 0, 0)}

    c_vals = [to_int[ch] for ch in word]
    total  = sum(c_vals)

    S0 = sum((i + 1) * v for i, v in enumerate(c_vals)) % (n + 1)
    P0 = total % M
    K  = {(n, S0, P0)}                       # distance‑0 key

    # ---------- substitutions ------------------------------------------
    for j, v_old in enumerate(c_vals):
        for v_new in range(1, sigma + 1):
            if v_new == v_old:
                continue
            diff = v_new - v_old
            K.add((n,
                   (S0 + diff * (j + 1)) % (n + 1),
                   (P0 + diff) % M))

    # ---------- deletions ----------------------------------------------
    for j in range(n):
        K.add(vt_key(word[:j] + word[j + 1:]))

    # ---------- insertions ---------------------------------------------
    for j in range(n + 1):
        for x in range(1, sigma + 1):
            K.add(vt_key(word[:j] + ALPHABET[x - 1] + word[j:]))

    # ---------- adjacent transpositions --------------------------------
    # swap indices j and j+1
    if n >= 2:
        for j in range(n - 1):
            swapped = (word[:j] + word[j + 1] + word[j] + word[j + 2:])
            K.add(vt_key(swapped))

    return K

# --- variants within *k* edits (k ≥ 1) ----------------------------------
def vt_variants_k(word: str, k: int):
    """
    Return the set of keys reachable from *word* by ≤ k
    Damerau‑Levenshtein edits.

    Warning: grows exponentially; practical for k ≤ 2.
    """
    seen_words = {word}
    frontier   = {word}
    keys       = set()

    for _ in range(k):
        next_frontier = set()
        for w in frontier:
            for key in vt_variants(w):     # distance‑1 keys of w
                keys.add(key)
            # enumerate actual strings at dist‑1 to feed next layer
            splits  = [(w[:i], w[i:]) for i in range(len(w) + 1)]
            deletes = [L + R[1:] for L, R in splits if R]
            repls   = [L + c + R[1:] for L, R in splits if R
                       for c in ALPHABET if c != R[0]]
            inserts = [L + c + R for L, R in splits for c in ALPHABET]
            transps = [w[:i] + w[i + 1] + w[i] + w[i + 2:]
                       for i in range(len(w) - 1)]
            for v in itertools.chain(deletes, repls, inserts, transps):
                if v not in seen_words:
                    seen_words.add(v)
                    next_frontier.add(v)
        frontier = next_frontier
        if not frontier:
            break

    # finally, include the original word’s key, too
    keys.add(vt_key(word))
    return keys

# --- empirical test -----------------------------------------------------
def bound(n): return sigma*(2*n+1) + n + 1

# OK = True
# for _ in range(200):                        # 200 random words, len 1–10
#     w     = ''.join(random.choice(ALPHABET) for _ in range(random.randint(1,10)))
#     k_alg = vt_variants(w)
#     k_true = {vt_key(v) for v in edits1_damerau(w)} | {vt_key(w)}

#     if k_true != k_alg:
#         print("❌ mismatch on", w)
#         OK = False
#         break

#     assert len(k_alg) <= bound(len(w)), "size bound violated"

# print("All tests passed ✅" if OK else "Test failed")

# # test k = 1  (should still pass)
# for _ in range(200):
#     w = ''.join(random.choice(ALPHABET) for _ in range(random.randint(1,10)))
#     assert {vt_key(v) for v in edits1(w)} | {vt_key(w)} | \
#            {vt_key(w[:i]+w[i+1]+w[i]+w[i+2:])         # swaps
#             for i in range(len(w)-1)} <= vt_variants(w)

# # test k = 2
# for _ in range(100):
#     w = ''.join(random.choice(ALPHABET) for _ in range(random.randint(1,8)))
#     keys_alg = vt_variants_k(w, 2)
#     # brute‑force all distance‑2 strings:
#     true_set = {vt_key(x)
#                 for v1 in edits1(w)
#                 for x  in edits1(v1)} | {vt_key(w)}
#     assert true_set <= keys_alg

# print("All expanded‑k tests passed ✅")